classifier.py

```diff
@@ -3130,59 +3090,436 @@ def target_age(cls, identifier, name):
+        #
+        # 1. The weight of that audience is more than twice the
+        # combined weight of the 'adult' and 'adults only' audiences.
+        #
+        # 2. The 'adult' and 'adults only' audiences have no weight
+        # whatsoever, and the weight of the proposed audience is
+        # greater than 10.
+        if total_adult_weight:
+            threshold = total_adult_weight * 2
+        else:
+            threshold = 10
+
+        # If both the 'children' weight and the 'YA' weight pass the
+        # threshold, we go with 'YA'.
+        if (children_weight > threshold and children_weight > ya_weight):
+            audience = Classifier.AUDIENCE_CHILDREN
+        elif ya_weight + children_weight > threshold:
+            audience = Classifier.AUDIENCE_YOUNG_ADULT
+
+        # If the 'adults only' weight is more than 1/4 of the total adult
+        # weight, classify as 'adults only' to be safe.
+        #
+        # TODO: This has not been calibrated.
+        if (audience==Classifier.AUDIENCE_ADULT
+            and adults_only_weight > total_adult_weight/4):
+            audience = Classifier.AUDIENCE_ADULTS_ONLY
+
+        return audience
+
+    def target_age(self, audience):
+        """Derive a target age from the gathered data."""
+
+        if audience not in (
+                Classifier.AUDIENCE_CHILDREN, Classifier.AUDIENCE_YOUNG_ADULT
+        ):
+            # This is not a children's or YA book. Assertions about
+            # target age are irrelevant and the default value rules.
+            return Classifier.default_target_age_for_audience(audience)
+
+        # Only consider the most reliable classifications.
+        reliable_classifications = self.most_reliable_target_age_subset
+
+        # Try to reach consensus on the lower and upper bounds of the
+        # age range.
+        target_age_mins = []
+        target_age_maxes = []
+        for c in reliable_classifications:
+            target_min = c.subject.target_age.lower
+            target_max = c.subject.target_age.upper
+            if target_min is not None:
+                if not c.subject.target_age.lower_inc:
+                    target_min += 1
+                for i in range(0,c.weight):
+                    target_age_mins.append(target_min)
+            if target_max is not None:
+                if not c.subject.target_age.upper_inc:
+                    target_max -= 1
+                for i in range(0,c.weight):
+                    target_age_maxes.append(target_max)
+
+        target_age_min = None
+        target_age_max = None
+        if target_age_mins:
+            # We have opinions about the proper minimum age.
+            mins = Counter(target_age_mins)
+            [(target_age_min, count)] = mins.most_common(1)
+            if count == 1:
+                # Everyone has a different opinion. Pick the smallest
+                # minimum age.
+                target_age_min = min(target_age_mins)
```

Comment:
Good question. I haven't really seen that come up but it is a possibility. I've refactored the code to take that into account.